### ISSUE TITLE
JSON-LD Mapping

### ISSUE THREAD

**Issue by denjello** [ito-suite/ng-ito-dataviewer#1]
We need to have a workflow that maps our API responses to json-ld (in a variety of flavors). It would be clever to enable the settings of this mapping on the ORG management page, and even allow the ORG manager to create their own mappings to RDF-a, microformat, vanilla - or even Dublin Core Initiative.

This last issue has been on the table since early 2014, and it is time to tackle it while we are redesigning the apis. It may however require that we bite the bullet and push out our own version of json-ld > DCI translator. :beers:

**Comment by dasantonym** [comment]
sounds cool but also super complex and a ui nightmare. a few questions that immediately come to mind: isn't this really error prone if some admin maps actually predefined structures? is there even any room for interpretations or should this rather be a preset? what happens if the admin changes the format on an existing archive? how did you add those beers? can i please have one?

**Comment by denjello** [comment]
No, the mappings do nothing to the actual structure of the DB. They are tacked on to the server response as a way of interpreting the data for "linked-data" purposes and is part of the JSON-LD spec. 

:beers: = `:beers:`

**Comment by dasantonym** [comment]
still not sure why this needs to be configurable. isn't there one way only to map it and then it's mapped and then why would you want to remap it differently (i'd rather put it in the source or config and version the mapping for everyone)? maybe i need to read up on this stuff first.

:beers: :beer: :bee:

**Comment by denjello** [comment]
in the short term it doesn't have to be configurable. however, I can foresee situations where the org users might like to be able to integrate the system key:values to specific mappings in microdata or dublin core - and I think that they are the ones to be doing the mappings (because I'll be dead before I do all of those!!!) 

But here is exactly what I am talking about in the spec:
http://www.w3.org/TR/json-ld/#the-context

**Comment by dasantonym** [comment]
ah yeah i see. but i'd really not want to go all the way and build an editor for this. this would be an ontology editor and it is a separate project on it's own. our ontology is static since the data structure does not change and you can not have an end user build this. that's like having the content manager build your db structure... but it is crucial if there is a linked data approach so i'd say this should go to the core system.

i mean, if you manage to build this then that's great obvs, but this will be far more complex than just doing the mappings in a json file or something and having them statically committed with the json-ld or the dc or the configuration module (probably best) or whatever.

**Comment by dasantonym** [comment]
dammit. tl:dr data structure and output formats are highly implementation specific. this will venture into drupal-ish "edit everything" territory and does not make sense to be edited on its own

**Comment by denjello** [comment]
Yeah - the drupal shit bugged me too. I am totally ok with doing it in config without GUI. 

Maybe we should open that repo and move this issue there.

**Comment by denjello** [comment]
i mean, i would generate a basic JSON-LD mapping that is useful for us, and if somebody needs to do it differently then it would be possible via the config system, but not something we need to actively support.

**Comment by dasantonym** [comment]
yes, that would be my favourite way to go! if the system is adapted to a totally different structure/use-case it will be minor to change the mappings compared to the rest of the changes anyway
